#### READ CSV FILE WITH PANDAS

In [1]:
from urllib.request import urlretrieve

In [2]:
italy_covid_url = 'https://gist.githubusercontent.com/aakashns/f6a004fa20c84fec53262f9a8bfee775/raw/f309558b1cf5103424cef58e2ecb8704dcd4d74c/italy-covid-daywise.csv'

urlretrieve(italy_covid_url, 'italy-covid-daywise.csv')

('italy-covid-daywise.csv', <http.client.HTTPMessage at 0x2769a669ac0>)

We can read CSV files using a special pandas method.

In [11]:
import sys
!{sys.executable} -m pip install pandas

  Using cached pytz-2024.1-py2.py3-none-any.whl.metadata (22 kB)
  Using cached tzdata-2024.1-py2.py3-none-any.whl.metadata (1.4 kB)
   ---------------------------------------- 0.0/11.5 MB ? eta -:--:--
   ---------------------------------------- 0.0/11.5 MB 660.6 kB/s eta 0:00:18
    --------------------------------------- 0.2/11.5 MB 2.1 MB/s eta 0:00:06
   ---- ----------------------------------- 1.3/11.5 MB 10.1 MB/s eta 0:00:02
   -------- ------------------------------- 2.6/11.5 MB 14.9 MB/s eta 0:00:01
   ---------------- ----------------------- 4.6/11.5 MB 21.0 MB/s eta 0:00:01
   ---------------------- ----------------- 6.6/11.5 MB 24.6 MB/s eta 0:00:01
   -------------------------- ------------- 7.5/11.5 MB 24.0 MB/s eta 0:00:01
   ---------------------------------- ----- 10.0/11.5 MB 27.9 MB/s eta 0:00:01
   ---------------------------------------  11.5/11.5 MB 36.4 MB/s eta 0:00:01
   ---------------------------------------  11.5/11.5 MB 36.4 MB/s eta 0:00:01
   -----------

In [12]:
import pandas as pd

In [15]:
covid_df = pd.read_csv('italy-covid-daywise.csv')
covid_df

date  new_cases  new_deaths  new_tests
0    2019-12-31        0.0         0.0        NaN
1    2020-01-01        0.0         0.0        NaN
2    2020-01-02        0.0         0.0        NaN
3    2020-01-03        0.0         0.0        NaN
4    2020-01-04        0.0         0.0        NaN
..          ...        ...         ...        ...
243  2020-08-30     1444.0         1.0    53541.0
244  2020-08-31     1365.0         4.0    42583.0
245  2020-09-01      996.0         6.0    54395.0
246  2020-09-02      975.0         8.0        NaN
247  2020-09-03     1326.0         6.0        NaN

[248 rows x 4 columns]

Data from a CSV file will be stored in a DataFrame object.

In [14]:
type(covid_df)

pandas.core.frame.DataFrame

Here's what the dataframe shows us:
- four daywise counts for Covid-19 in Italy
- Metrics reports are new cases, new deaths, and new tests
- 248 days worth of data (Dec 12, 2019 to Sep 3, 2020)

We can view basic information about a dataframe using .info(), .columns, .shape

In [16]:
covid_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 248 entries, 0 to 247
Data columns (total 4 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   date        248 non-null    object 
 1   new_cases   248 non-null    float64
 2   new_deaths  248 non-null    float64
 3   new_tests   135 non-null    float64
dtypes: float64(3), object(1)
memory usage: 7.9+ KB


In [18]:
covid_df.columns

Index(['date', 'new_cases', 'new_deaths', 'new_tests'], dtype='object')

In [19]:
covid_df.shape

(248, 4)

Each column contains a specific data type. For numeric columns, you can view statistical information (mean, standard deviation, min/max values, and number of non-empty values) using .describe()

In [17]:
covid_df.describe()

new_cases  new_deaths     new_tests
count   248.000000  248.000000    135.000000
mean   1094.818548  143.133065  31699.674074
std    1554.508002  227.105538  11622.209757
min    -148.000000  -31.000000   7841.000000
25%     123.000000    3.000000  25259.000000
50%     342.000000   17.000000  29545.000000
75%    1371.750000  175.250000  37711.000000
max    6557.000000  971.000000  95273.000000

#### RETRIEVE DATA FROM DATAFRAME
First, lets retrieve data from this dataframe. To do this, it helps to understand the internal representation of data in a dataframe. You can think of a dataframe as a dictionary of lists.

In [20]:
# Dataframe format is similar to this
covid_data_dict = {
    'data': ['2020-08-30', '2020-08-31'],
    'new_cases': [1444, 1365],
    'new_deaths': [1, 4],
    'new_tests': [53541, 42583],
}

This format has a few benefits:
- All values in a column have the same data type, so it's more efficient to store in a single array
- Retrieving values for a row requires simply indexing
- The representation is more compact (compared to the list of dictionaries we created using numpy where each dictionary is 1 row)

In [24]:
# Retrieving data using simply indexing
covid_data_dict['new_cases']

[1444, 1365]

In [25]:
covid_df['new_cases']

0         0.0
1         0.0
2         0.0
3         0.0
4         0.0
        ...  
243    1444.0
244    1365.0
245     996.0
246     975.0
247    1326.0
Name: new_cases, Length: 248, dtype: float64

Each column is represented using a data structure called a Series, which is basically a numpy array with extra methods/properties, so we can index it as well.

In [29]:
covid_df['new_cases'][246]

975.0

There's also an .at method that allows us to retrieve at a specific row/column.

In [30]:
covid_df.at[246, 'new_cases']

975.0

You can also pass a list of column names to access a subset of the dataframe.

Doing this creates a "view" of the original dataframe, which means that both variables will point to the same place in memory where this data is stored. This also means that altering the values in one variable will alter the values in the other variable. 

If you need an actual duplication of a dataframe, use .copy()

In [31]:
cases_df = covid_df[['date', 'new_cases']]
cases_df

date  new_cases
0    2019-12-31        0.0
1    2020-01-01        0.0
2    2020-01-02        0.0
3    2020-01-03        0.0
4    2020-01-04        0.0
..          ...        ...
243  2020-08-30     1444.0
244  2020-08-31     1365.0
245  2020-09-01      996.0
246  2020-09-02      975.0
247  2020-09-03     1326.0

[248 rows x 2 columns]

In [32]:
covid_df_copy = covid_df.copy()

To access values at a specific row:

In [33]:
covid_df.loc[243]

date          2020-08-30
new_cases         1444.0
new_deaths           1.0
new_tests        53541.0
Name: 243, dtype: object

In [34]:
type(covid_df.loc[243])

pandas.core.series.Series

To get the first or last few rows:

In [35]:
covid_df.head()

date  new_cases  new_deaths  new_tests
0  2019-12-31        0.0         0.0        NaN
1  2020-01-01        0.0         0.0        NaN
2  2020-01-02        0.0         0.0        NaN
3  2020-01-03        0.0         0.0        NaN
4  2020-01-04        0.0         0.0        NaN

In [36]:
covid_df.tail(3)

date  new_cases  new_deaths  new_tests
245  2020-09-01      996.0         6.0    54395.0
246  2020-09-02      975.0         8.0        NaN
247  2020-09-03     1326.0         6.0        NaN

Notice that some row/column locations are 0 or NaN. This is because there may be some missing data. 

0 and NaN are distinct - NaN means "not a number", indicating lack of information, while 0 is just that, 0.

We can find the first index in a dataframe that doesnt contain NaN using .first_valid_index()

In [43]:
# NaN datatype is a float
type(covid_df.at[0, 'new_tests'])

numpy.float64

In [41]:
covid_df.new_tests.first_valid_index()

111

To verify that first_valid_index() works:

In [53]:
covid_df.loc[108:113]

date  new_cases  new_deaths  new_tests
108  2020-04-17     3786.0       525.0        NaN
109  2020-04-18     3493.0       575.0        NaN
110  2020-04-19     3491.0       480.0        NaN
111  2020-04-20     3047.0       433.0     7841.0
112  2020-04-21     2256.0       454.0    28095.0
113  2020-04-22     2729.0       534.0    44248.0

The method .sample() can be used to retrieve a random sample from the dataframe.

In [54]:
covid_df.sample(10)

date  new_cases  new_deaths  new_tests
147  2020-05-26      300.0        92.0    33944.0
123  2020-05-02     1965.0       269.0    31231.0
201  2020-07-19      249.0        14.0    20621.0
210  2020-07-28      168.0         5.0    25341.0
6    2020-01-06        0.0         0.0        NaN
230  2020-08-17      477.0         4.0    21379.0
10   2020-01-10        0.0         0.0        NaN
148  2020-05-27      397.0        78.0    37299.0
199  2020-07-17      230.0        20.0    28661.0
231  2020-08-18      320.0         4.0    32687.0

When we take a random sample from a dataframe, the original index of each row is preserved. This is another useful feature of dataframes.

#### ANALYZE DATA FROM DATAFRAME

##### What is the total number of reported cases and deaths?

pandas has a .sum(), just like numpy, that totals all values in a series.

In [55]:
total_cases = covid_df.new_cases.sum()
total_deaths = covid_df.new_deaths.sum()

In [67]:
# a series is any sequence of information (row/column)
type(covid_df.new_cases)

pandas.core.series.Series

In [56]:
print(f'The number of reported cases is {total_cases} and the number of total deaths is {total_deaths}')

The number of reported cases is 271515.0 and the number of total deaths is 35497.0


##### What is the overall death ratio (ratio of reported deaths to reported cases)?

In [81]:
death_rate = (covid_df.new_deaths.sum() / covid_df.new_cases.sum()) * 100

In [82]:
print(f'The overall reported death rate in Italy is {death_rate :.2f}%')

The overall reported death rate in Italy is 13.07%


##### What is the overall number of tests conducted? A total of 935310 tests were conducted before tracking started.

In [76]:
initial_tests = 935310
total_tests = initial_tests + covid_df.new_tests.sum()
total_tests

5214766.0

##### What fraction of tests returned a positive result?

In [91]:
positive_rate = (total_cases / total_tests) * 100

In [92]:
print(f'{positive_rate :.2f}% of tests in Italy led to a positive diagnosis.')

5.21% of tests in Italy led to a positive diagnosis.


In [93]:
covid_df.columns

Index(['date', 'new_cases', 'new_deaths', 'new_tests'], dtype='object')

##### What fraction of new cases resulted in death?

In [94]:
death_rate = (total_deaths / total_cases) * 100

In [96]:
print(f'{death_rate :.2f}% of positive cases resulted in death in Italy.')

13.07% of positive cases resulted in death in Italy.


##### What was the average number of new cases recorded?

In [106]:
avg_new_cases = covid_df.new_cases.mean()

In [113]:
print(f'The average number of new cases per reporting in Italy was about {avg_new_cases :.0f}.')

The average number of new cases per reporting in Italy was about 1095.


In [112]:
covid_df.describe()

new_cases  new_deaths     new_tests
count   248.000000  248.000000    135.000000
mean   1094.818548  143.133065  31699.674074
std    1554.508002  227.105538  11622.209757
min    -148.000000  -31.000000   7841.000000
25%     123.000000    3.000000  25259.000000
50%     342.000000   17.000000  29545.000000
75%    1371.750000  175.250000  37711.000000
max    6557.000000  971.000000  95273.000000

#### QUERYING AND SORTING ROWS

We can use boolean expressions to filter our information.

In [114]:
high_new_cases = covid_df.new_cases > 1000
high_new_cases

0      False
1      False
2      False
3      False
4      False
       ...  
243     True
244     True
245    False
246    False
247     True
Name: new_cases, Length: 248, dtype: bool

In [115]:
covid_df[high_new_cases]

date  new_cases  new_deaths  new_tests
68   2020-03-08     1247.0        36.0        NaN
69   2020-03-09     1492.0       133.0        NaN
70   2020-03-10     1797.0        98.0        NaN
72   2020-03-12     2313.0       196.0        NaN
73   2020-03-13     2651.0       189.0        NaN
..          ...        ...         ...        ...
241  2020-08-28     1409.0         5.0    65135.0
242  2020-08-29     1460.0         9.0    64294.0
243  2020-08-30     1444.0         1.0    53541.0
244  2020-08-31     1365.0         4.0    42583.0
247  2020-09-03     1326.0         6.0        NaN

[72 rows x 4 columns]

We can simplify our code like so:

In [117]:
high_new_cases = covid_df[covid_df.new_cases > 1000]
high_new_cases

date  new_cases  new_deaths  new_tests
68   2020-03-08     1247.0        36.0        NaN
69   2020-03-09     1492.0       133.0        NaN
70   2020-03-10     1797.0        98.0        NaN
72   2020-03-12     2313.0       196.0        NaN
73   2020-03-13     2651.0       189.0        NaN
..          ...        ...         ...        ...
241  2020-08-28     1409.0         5.0    65135.0
242  2020-08-29     1460.0         9.0    64294.0
243  2020-08-30     1444.0         1.0    53541.0
244  2020-08-31     1365.0         4.0    42583.0
247  2020-09-03     1326.0         6.0        NaN

[72 rows x 4 columns]

We can formulate more complex queries using multiple columns: try to find the days the ratio of cases reported to tests conducted is higher than the overall positive rate.

In [124]:
high_ratio_df = covid_df[(covid_df.new_cases / covid_df.new_tests) * 100 > positive_rate]
high_ratio_df

date  new_cases  new_deaths  new_tests
111  2020-04-20     3047.0       433.0     7841.0
112  2020-04-21     2256.0       454.0    28095.0
113  2020-04-22     2729.0       534.0    44248.0
114  2020-04-23     3370.0       437.0    37083.0
116  2020-04-25     3021.0       420.0    38676.0
117  2020-04-26     2357.0       415.0    24113.0
118  2020-04-27     2324.0       260.0    26678.0
120  2020-04-29     2091.0       382.0    38589.0
123  2020-05-02     1965.0       269.0    31231.0
124  2020-05-03     1900.0       474.0    27047.0
125  2020-05-04     1389.0       174.0    22999.0
128  2020-05-07     1444.0       369.0    13665.0

You can add a new column to a dataframe (similar to how you would adding a key to a dictionary).

In [127]:
covid_df['positive_rate'] = (covid_df.new_cases / covid_df.new_tests) * 100
covid_df

date  new_cases  new_deaths  new_tests  positive_rate
0    2019-12-31        0.0         0.0        NaN            NaN
1    2020-01-01        0.0         0.0        NaN            NaN
2    2020-01-02        0.0         0.0        NaN            NaN
3    2020-01-03        0.0         0.0        NaN            NaN
4    2020-01-04        0.0         0.0        NaN            NaN
..          ...        ...         ...        ...            ...
243  2020-08-30     1444.0         1.0    53541.0       2.696999
244  2020-08-31     1365.0         4.0    42583.0       3.205505
245  2020-09-01      996.0         6.0    54395.0       1.831051
246  2020-09-02      975.0         8.0        NaN            NaN
247  2020-09-03     1326.0         6.0        NaN            NaN

[248 rows x 5 columns]

Consider the context of this dataset: In real life, it would take a few days to get the results of a test, so we can't actually compare the information. positive_rate would likely contain inaccurate data, so we'll drop it.

In [128]:
covid_df.drop(columns=['positive_rate'], inplace=True)

The inplace argument allows us to change some data at its source (it's position in memory). 

##### Sorting rows
Rows can be sorted by column using .sort_values(). Lets sort to identify the days with the highest number of cases, then chain it with the head method. Then, lets sort to find the days with the highest number of deaths.

In [131]:
highest_10_cases = covid_df.sort_values('new_cases', ascending=False).head(10)
highest_10_cases

date  new_cases  new_deaths  new_tests
82  2020-03-22     6557.0       795.0        NaN
87  2020-03-27     6153.0       660.0        NaN
81  2020-03-21     5986.0       625.0        NaN
89  2020-03-29     5974.0       887.0        NaN
88  2020-03-28     5959.0       971.0        NaN
83  2020-03-23     5560.0       649.0        NaN
80  2020-03-20     5322.0       429.0        NaN
85  2020-03-25     5249.0       743.0        NaN
90  2020-03-30     5217.0       758.0        NaN
86  2020-03-26     5210.0       685.0        NaN

In [132]:
highest_10_deaths = covid_df.sort_values('new_deaths', ascending=False).head(10)
highest_10_deaths

date  new_cases  new_deaths  new_tests
88  2020-03-28     5959.0       971.0        NaN
89  2020-03-29     5974.0       887.0        NaN
92  2020-04-01     4053.0       839.0        NaN
91  2020-03-31     4050.0       810.0        NaN
82  2020-03-22     6557.0       795.0        NaN
95  2020-04-04     4585.0       764.0        NaN
94  2020-04-03     4668.0       760.0        NaN
90  2020-03-30     5217.0       758.0        NaN
85  2020-03-25     5249.0       743.0        NaN
93  2020-04-02     4782.0       727.0        NaN

The last 2 weeks of March showed the highest cases. Between the end of March and the beginning of April is when we see the highest number of deaths.

Lets look at the days with the least number of cases:

In [133]:
least_10_cases = covid_df.sort_values('new_cases', ascending=True).head(10)
least_10_cases

date  new_cases  new_deaths  new_tests
172  2020-06-20     -148.0        47.0    29875.0
0    2019-12-31        0.0         0.0        NaN
29   2020-01-29        0.0         0.0        NaN
30   2020-01-30        0.0         0.0        NaN
32   2020-02-01        0.0         0.0        NaN
33   2020-02-02        0.0         0.0        NaN
34   2020-02-03        0.0         0.0        NaN
36   2020-02-05        0.0         0.0        NaN
37   2020-02-06        0.0         0.0        NaN
38   2020-02-07        0.0         0.0        NaN

We could've predicted that the entries on this list would occur around the beginning of that year, but the data shows a negative value as the top entry at a date in June. This can't be right, but we can infer that this may be due to data entry error or attempts to correct miscounting.

This is the nature of real world data, so it's useful to find context.

In [134]:
covid_df.loc[168:175]

date  new_cases  new_deaths  new_tests
168  2020-06-16      301.0        26.0    27762.0
169  2020-06-17      210.0        34.0    33957.0
170  2020-06-18      328.0        43.0    32921.0
171  2020-06-19      331.0        66.0    28570.0
172  2020-06-20     -148.0        47.0    29875.0
173  2020-06-21      264.0        49.0    24581.0
174  2020-06-22      224.0        24.0    16152.0
175  2020-06-23      221.0        23.0    23225.0

If this was a data entry error, we can choose from the following options:
- replace it with 0
- replace it with the average of the entire column
- replace it with the average of the values on the previous and next date
- discard the row

Your choice should depend on context about the data and the problem. Since we are dealing with date-ordered data, we'll go with the third option.

We can use the .at method for this.

In [136]:
covid_df.at[172, 'new_cases'] = (covid_df.at[171, 'new_cases'] + 
                                 covid_df.at[173, 'new_cases']) / 2
covid_df.at[172, 'new_cases']

297.5

##### Working with dates

Lets study this data on a monthly basis. pandas has built-in functions for dates.

In [137]:
covid_df.date

0      2019-12-31
1      2020-01-01
2      2020-01-02
3      2020-01-03
4      2020-01-04
          ...    
243    2020-08-30
244    2020-08-31
245    2020-09-01
246    2020-09-02
247    2020-09-03
Name: date, Length: 248, dtype: object

The dtype of this tells us that pandas doesnt yet know this column is filled with dates. We can specify it as a datetime column using pd.to_datetime

In [141]:
covid_df['date'] = pd.to_datetime(covid_df.date)
covid_df['date']

0     2019-12-31
1     2020-01-01
2     2020-01-02
3     2020-01-03
4     2020-01-04
         ...    
243   2020-08-30
244   2020-08-31
245   2020-09-01
246   2020-09-02
247   2020-09-03
Name: date, Length: 248, dtype: datetime64[ns]

Now dtype is datetime64. We can extract parts of this datetime data using the DatetimeIndex class.

In [142]:
covid_df['year'] = pd.DatetimeIndex(covid_df.date).year
covid_df['month'] = pd.DatetimeIndex(covid_df.date).month
covid_df['day'] = pd.DatetimeIndex(covid_df.date).day
covid_df['weekday'] = pd.DatetimeIndex(covid_df.date).weekday

In [143]:
covid_df

date  new_cases  new_deaths  new_tests  year  month  day  weekday
0   2019-12-31        0.0         0.0        NaN  2019     12   31        1
1   2020-01-01        0.0         0.0        NaN  2020      1    1        2
2   2020-01-02        0.0         0.0        NaN  2020      1    2        3
3   2020-01-03        0.0         0.0        NaN  2020      1    3        4
4   2020-01-04        0.0         0.0        NaN  2020      1    4        5
..         ...        ...         ...        ...   ...    ...  ...      ...
243 2020-08-30     1444.0         1.0    53541.0  2020      8   30        6
244 2020-08-31     1365.0         4.0    42583.0  2020      8   31        0
245 2020-09-01      996.0         6.0    54395.0  2020      9    1        1
246 2020-09-02      975.0         8.0        NaN  2020      9    2        2
247 2020-09-03     1326.0         6.0        NaN  2020      9    3        3

[248 rows x 8 columns]

Lets check overall metrics for the month of May. We can query the rows for May, choose a subset of column that we want to aggregate, and use the sum method.

In [145]:
# Query the rows for May
covid_df_may = covid_df[covid_df.month == 5]

# Extract the subset of columns
covid_df_may_metrics = covid_df_may[['new_cases', 'new_deaths', 'new_tests']]

# Get the column-wise sum
covid_may_totals = covid_df_may_metrics.sum()

In [146]:
covid_df_may

date  new_cases  new_deaths  new_tests  year  month  day  weekday
122 2020-05-01     1872.0       285.0    43732.0  2020      5    1        4
123 2020-05-02     1965.0       269.0    31231.0  2020      5    2        5
124 2020-05-03     1900.0       474.0    27047.0  2020      5    3        6
125 2020-05-04     1389.0       174.0    22999.0  2020      5    4        0
126 2020-05-05     1221.0       195.0    32211.0  2020      5    5        1
127 2020-05-06     1075.0       236.0    37771.0  2020      5    6        2
128 2020-05-07     1444.0       369.0    13665.0  2020      5    7        3
129 2020-05-08     1401.0       274.0    45428.0  2020      5    8        4
130 2020-05-09     1327.0       243.0    36091.0  2020      5    9        5
131 2020-05-10     1083.0       194.0    31384.0  2020      5   10        6
132 2020-05-11      802.0       165.0    25823.0  2020      5   11        0
133 2020-05-12      744.0       179.0    39620.0  2020      5   12        1
134 2020-05-13     1402.0       172.0    37049.0  2020      5   13        2
135 2020-05-14      888.0       195.0    41131.0  2020      5   14        3
136 2020-05-15      992.0       262.0    39027.0  2020      5   15        4
137 2020-05-16      789.0       242.0    40657.0  2020      5   16        5
138 2020-05-17      875.0       153.0    33505.0  2020      5   17        6
139 2020-05-18      675.0       145.0    26101.0  2020      5   18        0
140 2020-05-19      451.0        99.0    40226.0  2020      5   19        1
141 2020-05-20      813.0       162.0    38617.0  2020      5   20        2
142 2020-05-21      665.0       161.0    40644.0  2020      5   21        3
143 2020-05-22      642.0       156.0    42987.0  2020      5   22        4
144 2020-05-23      652.0       130.0    42579.0  2020      5   23        5
145 2020-05-24      669.0       119.0    34206.0  2020      5   24        6
146 2020-05-25      531.0        50.0    20676.0  2020      5   25        0
147 2020-05-26      300.0        92.0    33944.0  2020      5   26        1
148 2020-05-27      397.0        78.0    37299.0  2020      5   27        2
149 2020-05-28      584.0       117.0    39838.0  2020      5   28        3
150 2020-05-29      593.0        70.0    38233.0  2020      5   29        4
151 2020-05-30      516.0        87.0    36051.0  2020      5   30        5
152 2020-05-31      416.0       111.0    28948.0  2020      5   31        6

In [147]:
covid_df_may_metrics

new_cases  new_deaths  new_tests
122     1872.0       285.0    43732.0
123     1965.0       269.0    31231.0
124     1900.0       474.0    27047.0
125     1389.0       174.0    22999.0
126     1221.0       195.0    32211.0
127     1075.0       236.0    37771.0
128     1444.0       369.0    13665.0
129     1401.0       274.0    45428.0
130     1327.0       243.0    36091.0
131     1083.0       194.0    31384.0
132      802.0       165.0    25823.0
133      744.0       179.0    39620.0
134     1402.0       172.0    37049.0
135      888.0       195.0    41131.0
136      992.0       262.0    39027.0
137      789.0       242.0    40657.0
138      875.0       153.0    33505.0
139      675.0       145.0    26101.0
140      451.0        99.0    40226.0
141      813.0       162.0    38617.0
142      665.0       161.0    40644.0
143      642.0       156.0    42987.0
144      652.0       130.0    42579.0
145      669.0       119.0    34206.0
146      531.0        50.0    20676.0
147      300.0        92.0    33944.0
148      397.0        78.0    37299.0
149      584.0       117.0    39838.0
150      593.0        70.0    38233.0
151      516.0        87.0    36051.0
152      416.0       111.0    28948.0

In [148]:
covid_may_totals

new_cases       29073.0
new_deaths       5658.0
new_tests     1078720.0
dtype: float64

In [149]:
type(covid_may_totals)

pandas.core.series.Series

We can also combine this code into a single line as:

In [151]:
covid_df[covid_df.month == 5][['new_cases', 'new_deaths', 'new_tests']].sum()

new_cases       29073.0
new_deaths       5658.0
new_tests     1078720.0
dtype: float64

Lets check if the number of cases reported on Sundays is higher than the average number of cases per day. This requires the .mean method

In [152]:
# Overall average
covid_df.new_cases.mean()

1096.6149193548388

In [153]:
# Average for Sundays
covid_df[covid_df.weekday == 6].new_cases.mean()

1247.2571428571428

The average cases reported on Sundays is higher than the overall average of cases reported.

#### GROUPING AND AGGREGATION

Next, lets summarize daywise data and create a new dataframe with month-wise data. We'll use the .groupby() function - specify the grouping and way to aggregate data for each group.

In [156]:
covid_month_df = covid_df.groupby('month')[['new_cases', 'new_deaths', 'new_tests']].sum()

In [157]:
covid_month_df

new_cases  new_deaths  new_tests
month                                  
1            3.0         0.0        0.0
2          885.0        21.0        0.0
3       100851.0     11570.0        0.0
4       101852.0     16091.0   419591.0
5        29073.0      5658.0  1078720.0
6         8217.5      1404.0   830354.0
7         6722.0       388.0   797692.0
8        21060.0       345.0  1098704.0
9         3297.0        20.0    54395.0
12           0.0         0.0        0.0

## EVERY TOOL FROM THIS SECTION:

- import sys
- !{sys.executable}

Reading CSV files:
- pd.read_csv()
- .info()
- .describe()
- .columns
- .shape

Retrieving data:
- covid_df['new_cases']
- new_cases[243]
- covid_df.at[243, 'new_cases']
- covid_df.copy()
- covid_df.loc[243]
- .head(), .tail(), .sample()
- covid_df.new_tests.first_valid_index

Analyzing data:
- .sum()
- .mean()
- .describe()

Querying and sorting rows:
- conditional statements
- adding columns to dataframes
- .drop()
- .sort_values()
- replacing values using .at
- .date
- pd.to_datetime()
- pd.DatetimeIndex()

Grouping and aggregation:

Merging data:

Basic plotting:
